app/src/main/java/se/nullable/flickboard/KeyboardService.kt

```diff
@@ -402,11 +404,26 @@ class KeyboardService : InputMethodService(), LifecycleOwner, SavedStateRegistry
 
                                 Action.ToggleEmojiMode ->
                                     emojiMode = !emojiMode
+
+                                Action.EnableVoiceMode -> {
+                                    voiceMode = true
+                                }
                             }
                         }
                         Box {
                             when {
                                 emojiMode -> EmojiKeyboard(onAction = onAction)
+                                voiceMode -> {
+                                    voiceMode = false
+                                    val voiceMethodId = getAvailableVoiceInputMethod()
+
+                                    if(voiceMethodId != null) {
+                                        // Only works on Android 9+
+                                        switchInputMethod((voiceMethodId))
+                                    } else {
+                                        // TODO - some sort of user feedback indicating no voice found
```

Comment:
I know in an earlier version that there was a debug popup message of some kind that would appear while typing on the Preview Keyboard; if that's an acceptable solution and you can point me to the method, I'm happy to hash this out.

Reply:
That was using a [Snackbar](https://developer.android.com/jetpack/compose/components/snackbar).